### DISCUSSION TITLE
Making a simple neural network model to detect handwritten digits

### DISCUSSION THREAD

**Discussion by Trollofjom0215** [community/community#60276] [Programming Help]
This is one of my first models, as MNIST dataset is fit for beginners I tried making a simple neural network model and improving on it step by step.
One of the doubts i'm having is whether my model is overfitting or not
I've imported the data, split it into train and test , normalized it, reshaped it according to need (apologies for my messy code-im just learning), and im trying to learn hyperparameter tuning. 
ive used the kerastuner library and got the following results :

code : 
import tensorflow as tf
(x_train, y_train),(x_test, y_test) = tf.keras.datasets.mnist.load_data()
print(x_train.shape)
print(y_train.shape)
(60000, 28, 28)
(60000,)
print(x_test.shape)
print(y_test.shape)
(10000, 28, 28)
(10000,)
*Dataset Statistics*
import numpy as np
print("Number of images : ", len(x_train))
print("Dimension of images (width,height) : ", x_train.shape[1:])
print("Number of classes : ", len(np.unique(y_train)))
Number of images :  60000
Dimension of images (width,height) :  (28, 28)
Number of classes :  10
*Data Visualization*
import matplotlib.pyplot as plt
# Visualize a few sample images
fig, axes = plt.subplots(nrows=2, ncols=5, figsize=(10, 5))
for i, ax in enumerate(axes.flat):
    ax.imshow(x_train[i] , cmap='gray')
    ax.set_title(f"Label: {y_train[i]}")
    ax.axis('off')

plt.tight_layout()
plt.show()

*Class Distribution*
class_counts = np.bincount(y_train)
print("counts/frequencies of different labels",class_counts)
print("shape of class_counts", class_counts.shape)
print("type of class counts : ",type(class_counts))
counts/frequencies of different labels [5923 6742 5958 6131 5842 5421 5918 6265 5851 5949]
shape of class_counts (10,)
type of class counts :  <class 'numpy.ndarray'>
classes = range(len(class_counts))

# Plot the class frequencies
plt.bar(classes, class_counts)
plt.xlabel("Class")
plt.ylabel("Frequency")
plt.title("Class Distribution in MNIST Dataset")
plt.xticks(classes) #sets x-axis tick  marks to match the class labels
plt.show()

*Data Preprocessing*
As we can see, data has no class imbalance, no noisy data, no outliers. Henceforth we just need to normalize the data and flatten it for our neural network to take in as input.
#converting x_train to 2-D array to apply MinMaxScaling(normalization)
num_samples, width, height = x_train.shape
x_train_reshaped = x_train.reshape(num_samples, width*height)
print(type(x_train), x_train.shape)
print(type(x_train_reshaped),x_train_reshaped.shape)
print(type(y_train),y_train.shape)
<class 'numpy.ndarray'> (60000, 28, 28)
<class 'numpy.ndarray'> (60000, 784)
<class 'numpy.ndarray'> (60000,)
y_train
array([5, 0, 4, ..., 5, 6, 8], dtype=uint8)
from sklearn.preprocessing import MinMaxScaler
scaler = MinMaxScaler()
scaler.fit(x_train_reshaped)
x_train_normalized = scaler.transform(x_train_reshaped)
x_train_normalized.shape
(60000, 784)
x_train_normalized.shape
(60000, 784)
y_train.shape
(60000,)
*HyperParameter Tuning*
import keras_tuner as kt
from keras.layers import Dropout
from keras.layers import Dense
from keras.models import Sequential 
def build_model(hp):
    model = Sequential()
    num_layers = hp.Int("num_layers" , min_value = 1, max_value = 10)
    activation = hp.Choice("activation", values = ['relu', 'sigmoid', 'tanh'])
    optimizer = hp.Choice("optimizer", values = ['rmsprop', 'adam', 'adagrad', 'adadelta'])
    for i in range(num_layers):
        model.add(Dense(units = hp.Int(f"units_{i}", min_value =16 ,max_value =128, step = 8), activation = activation, input_shape=(784,)))
    if hp.Boolean("dropout"):
        model.add(Dropout(rate = hp.Choice("rate", values = [0.2,0.4,0.6,0.8])))
    model.add(Dense(units=10, activation = 'softmax'))
    model.compile(optimizer=optimizer, loss= 'sparse_categorical_crossentropy', metrics = ['accuracy'] )
    
    return model
tuner = kt.RandomSearch(hypermodel= build_model , objective = 'val_accuracy', max_trials = 10, directory = 'mydir1', project_name = 'hyperparametertuning2')
tuner.search(x_train_normalized,y_train, epochs = 5 , validation_data = (x_test_2d,y_test))
Trial 10 Complete [00h 00m 26s]
val_accuracy: 0.11349999904632568

Best val_accuracy So Far: 0.9724000096321106
Total elapsed time: 00h 03m 58s
INFO:tensorflow:Oracle triggered exit
tuner.get_best_hyperparameters()[0].values
{'num_layers': 3,
 'activation': 'relu',
 'optimizer': 'adam',
 'units_0': 56,
 'dropout': False,
 'units_1': 88,
 'units_2': 16,
 'units_3': 120,
 'units_4': 32,
 'units_5': 80,
 'units_6': 104,
 'units_7': 112,
 'units_8': 120,
 'rate': 0.8,
 'units_9': 104}
num_layers = tuner.get_best_hyperparameters()[0].values['num_layers']
units = [
    tuner.get_best_hyperparameters()[0].values[f"units_{i}"]
    for i in range(num_layers)]
print(units)
[56, 88, 16]
model = tuner.get_best_models(num_models=1)[0]
model.summary()
Model: "sequential"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
=================================================================
 dense (Dense)               (None, 56)                43960     
                                                                 
 dense_1 (Dense)             (None, 88)                5016      
                                                                 
 dense_2 (Dense)             (None, 16)                1424      
                                                                 
 dense_3 (Dense)             (None, 10)                170       
                                                                 
=================================================================
Total params: 50,570
Trainable params: 50,570
Non-trainable params: 0
_________________________________________________________________
model.fit(x_train_normalized,y_train, epochs = 10, initial_epoch = 5, validation_data=(x_test_2d,y_test) )
Epoch 6/10
1875/1875 [==============================] - 3s 2ms/step - loss: 0.0173 - accuracy: 0.9942 - val_loss: 33.0315 - val_accuracy: 0.9732
Epoch 7/10
1875/1875 [==============================] - 4s 2ms/step - loss: 0.0152 - accuracy: 0.9949 - val_loss: 29.8919 - val_accuracy: 0.9744
Epoch 8/10
1875/1875 [==============================] - 4s 2ms/step - loss: 0.0159 - accuracy: 0.9945 - val_loss: 29.5549 - val_accuracy: 0.9763
Epoch 9/10
1875/1875 [==============================] - 4s 2ms/step - loss: 0.0129 - accuracy: 0.9955 - val_loss: 32.5522 - val_accuracy: 0.9750
Epoch 10/10
1875/1875 [==============================] - 4s 2ms/step - loss: 0.0151 - accuracy: 0.9949 - val_loss: 28.9002 - val_accuracy: 0.9751

**Comment by Anant-mishra1729** [comment]
It appears that your model is not overfitting. Here's why:

Training loss and accuracy: The training loss is relatively low (0.0151), and the training accuracy is high (0.9949). These values indicate that the model is performing well on the training data.

Validation loss and accuracy: The validation loss (28.9002) and validation accuracy (0.9751) are slightly worse than the training metrics, but they are still relatively good. The validation accuracy is consistently high across epochs, which suggests that the model is generalizing well to unseen data.
